torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -137,74 +137,75 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::
   return pythonRpc ? std::move(pythonRpc) : std::move(rpc);
 }
 
-void RequestCallbackImpl::processScriptCall(
-    RpcCommandBase& rpc,
-    const std::function<void(Message)>& markComplete,
-    const c10::intrusive_ptr<JitFuture>& responseFuture) const {
+c10::intrusive_ptr<JitFuture> RequestCallbackImpl::processScriptCall(
+    RpcCommandBase& rpc) const {
   auto& scriptCall = static_cast<ScriptCall&>(rpc);
-  auto& stack = scriptCall.stackRef();
+
+  c10::intrusive_ptr<JitFuture> future;
   if (scriptCall.hasOp()) {
-    processScriptCallOp(scriptCall, markComplete, stack);
-    return;
+    future = runJitOperator(*scriptCall.op(), scriptCall.stackRef());
+  } else {
+    future = runJitFunction(
+        scriptCall.qualifiedName(),
+        scriptCall.stackRef(),
+        scriptCall.isAsyncExecution());
   }
 
-  auto jitFuture = runJitFunction(
-      scriptCall.qualifiedName(), stack, scriptCall.isAsyncExecution());
-
-  jitFuture->addCallback([responseFuture, markComplete](JitFuture& jitFuture) {
-    if (jitFuture.hasError()) {
-      responseFuture->setError(jitFuture.exception_ptr());
-    } else {
-      responseFuture->markCompleted(c10::make_intrusive<Message>(
-          ScriptResp(jitFuture.value()).toMessage()));
-    }
-  });
+  return future->then(
+      [](JitFuture& jitFuture) {
+        return c10::make_intrusive<Message>(
+            ScriptResp(jitFuture.value()).toMessage());
+      },
+      c10::getCustomClassType<c10::intrusive_ptr<Message>>());
 }
 
-void RequestCallbackImpl::processPythonCall(
-    RpcCommandBase& rpc,
-    const std::function<void(Message)>& markComplete,
-    const c10::intrusive_ptr<JitFuture>& responseFuture) const {
+c10::intrusive_ptr<JitFuture> RequestCallbackImpl::processPythonCall(
+    RpcCommandBase& rpc) const {
   auto& upc = static_cast<UnpickledPythonCall&>(rpc);
   auto future = runPythonFunction(upc.pythonUdf(), upc.isAsyncExecution());
```

Comment:
I assume, later, we will need to pass agent devices to this Future?

Reply:
Correct, I'm doing it in #58426.